File: topological.py
from digraph import Digraph


class Topological:
    def __init__(self, g):
        self.marked = {}
        self.verts = []
        for v in g.getVerts():
            if v not in self.marked:
                self.__dfs(g, v)

    def getTopological(self):
        return self.verts

    def __dfs(self, g, s):
        self.marked[s] = True
        for w in g.getAdj(s):
            if w not in self.marked:
                self.__dfs(g, w)
        self.verts.insert(0, s)


if __name__ == "__main__":

    g = Digraph("tinydag.txt")

    topo = Topological(g)

    for v in topo.getTopological():
        print(f"{v} ", end="")
    print()

File: topological.cpp
#include "topological.h"
#include <iostream>
#include <fstream>
#include <sstream>

using namespace std;

Topological::Topological(Graph &g)
{
	for (auto const &v : g.getVerts()) {
    if (marked.find(v) == marked.end())
      dfs(g, v);
  }
}

void Topological::dfs(Graph &g, string v)
{
	marked.insert(v); // marca v como visitado
	for (auto const &w : g.getAdj(v))
	{
		if (marked.find(w) == marked.end())
			dfs(g, w);
	}
  topo.push_front(v);
}

const deque<string>& Topological::getTopological()
{
  return topo;
}
